src/libostree/ostree-repo.c

```diff
@@ -347,49 +347,56 @@ push_repo_lock (OstreeRepo          *self,
     {
       g_debug ("Creating repo lock table");
       lock_table = g_hash_table_new_full (NULL, NULL, NULL,
-                                          (GDestroyNotify)free_repo_lock);
+                                          (GDestroyNotify)g_ptr_array_unref);
       g_private_set (&repo_lock_table, lock_table);
     }
 
-  OstreeRepoLock *lock = g_hash_table_lookup (lock_table, self);
-  if (lock == NULL)
+  GPtrArray *stack = g_hash_table_lookup (lock_table, self);
+  if (stack == NULL)
+    {
+      stack = g_ptr_array_new ();
+      g_hash_table_insert (lock_table, self, stack);
+    }
+
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&self->lock.mutex);
+
+  if (self->lock.fd == -1)
     {
-      lock = g_new0 (OstreeRepoLock, 1);
-      g_queue_init (&lock->stack);
       g_debug ("Opening repo lock file");
-      lock->fd = TEMP_FAILURE_RETRY (openat (self->repo_dir_fd, ".lock",
-                                             O_CREAT | O_RDWR | O_CLOEXEC,
-                                             DEFAULT_REGFILE_MODE));
-      if (lock->fd < 0)
-        {
-          free_repo_lock (lock);
-          return glnx_throw_errno_prefix (error,
-                                          "Opening lock file %s/.lock failed",
-                                          gs_file_get_path_cached (self->repodir));
-        }
-      g_hash_table_insert (lock_table, self, lock);
+      self->lock.fd = TEMP_FAILURE_RETRY (openat (self->repo_dir_fd, ".lock",
+                                                  O_CREAT | O_RDWR | O_CLOEXEC,
+                                                  DEFAULT_REGFILE_MODE));
+      if (self->lock.fd < 0)
+        return glnx_throw_errno_prefix (error,
+                                        "Opening lock file %s/.lock failed",
+                                        gs_file_get_path_cached (self->repodir));
     }
 
   OstreeRepoLockInfo info;
-  repo_lock_info (lock, &info);
+  repo_lock_info (self, &info);
   g_debug ("Push lock: state=%s, depth=%u", info.name, info.len);
 
   if (info.state == LOCK_EX)
     {
       g_debug ("Repo already locked exclusively, extending stack");
-      g_queue_push_head (&lock->stack, GINT_TO_POINTER (LOCK_EX));
+      self->lock.exclusive++;
+      g_ptr_array_add (stack, GINT_TO_POINTER (LOCK_EX));
     }
   else
     {
       int next_state = (flags & LOCK_EX) ? LOCK_EX : LOCK_SH;
       const char *next_state_name = (flags & LOCK_EX) ? "exclusive" : "shared";
 
       g_debug ("Locking repo %s", next_state_name);
-      if (!do_repo_lock (lock->fd, flags))
+      if (!do_repo_lock (self->lock.fd, flags))
         return glnx_throw_errno_prefix (error, "Locking repo %s failed",
                                         next_state_name);
 
-      g_queue_push_head (&lock->stack, GINT_TO_POINTER (next_state));
+      if (next_state == LOCK_EX)
+        self->lock.exclusive++;
+      else
```

Comment:
I think both of these can overflow (wrap back to 0), breaking lock correctness.

Reply:
Right.